connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java

```
@@ -0,0 +1,578 @@
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.transforms.field;
+
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Schema.Type;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.transforms.util.SchemaUtil;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Multiple field paths to access data objects ({@code Struct} or {@code Map}) efficiently,
+ * instead of using {@see SingleFieldPath} individually.
+ * <p>
+ * If the SMT requires accessing a single field on the same data object,
+ * use {@code FieldPath} instead.
+ * <p>
+ * Invariants:
+ * <li>
+ *     <ul>Tree values contain either a nested tree or a field path</ul>
+ *     <ul>A tree cannot contain paths that are a subset of other paths (e.g. foo and foo.bar in V2 should collide and fail)</ul>
+ * </li>
+ *
+ * See KIP-821.
+ *
+ * @see SingleFieldPath
+ * @see FieldSyntaxVersion
+ */
+public class MultiFieldPaths implements FieldPath {
+
+    final Map<String, Object> pathTree;
+    final List<SingleFieldPath> paths;
+
+    MultiFieldPaths(List<SingleFieldPath> paths) {
+        this.paths = paths.stream().filter(Objects::nonNull).collect(Collectors.toList());
```

I'd err on the side of caution here and throw an exception if we see any null elements instead of silently filtering them out.

It's fairly safe to throw from SMTs given the `errors.tolerance` property and the options it gives users on how to handle failures while applying transformations.